sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -55,9 +55,9 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
       GpuRowToColumnarExec(optimizeAdaptiveTransitions(r2c.child, Some(r2c)), goal)
 
     case ColumnarToRowExec(GpuBringBackToHost(
-        GpuCoalesceBatches(e: GpuShuffleExchangeExecBase, _))) if parent.isEmpty =>
+        ShuffleCoalesceExec(e: GpuShuffleExchangeExecBase, _))) if parent.isEmpty =>
```

Comment:
Since `ShuffleCoalesceExec` is only inserted for legacy shuffle, don't we still need to handle the `GpuCoalesceBatches` case as well, for the RapidsShuffle case?

I am going to pull this PR locally and run some benchmarks with AQE and legacy shuffle but it would be good to do the same with UCX as well.

Reply:
Yes, nice catch!  This needs to insert the correct coalesce depending on the shuffle being used.

Reply:
Ok, so one way to handle this would be to duplicate the case statements, although we could simplify this if there is a common base class or trait.

```scala
    case ColumnarToRowExec(GpuBringBackToHost(
        ShuffleCoalesceExec(e: GpuShuffleExchangeExecBase, _))) if parent.isEmpty =>
      e.withNewChildren(e.children.map(c => optimizeAdaptiveTransitions(c, Some(e))))
    case ColumnarToRowExec(GpuBringBackToHost(
        GpuCoalesceBatches(e: GpuShuffleExchangeExecBase, _))) if parent.isEmpty =>
      e.withNewChildren(e.children.map(c => optimizeAdaptiveTransitions(c, Some(e))))
```

and then

```scala
            // Directly wrap shuffle query stage with coalesce batches operator
            if (GpuShuffleEnv.isRapidsShuffleEnabled) {
              GpuCoalesceBatches(s, TargetSize(conf.gpuTargetBatchSizeBytes))
            } else {
              ShuffleCoalesceExec(s, conf)
            }
```